ARCs/arc-0052.md

```diff
@@ -0,0 +1,210 @@
+---
+arc: 52
+title: Wallet Contextual KeyGen and Signing
+description: Algorand Wallet API for Contextual KeyGen and Signing
+author: Bruno Martins (@ehanoc), Patrick Bennett (@pbennett), Kevin Wellenzohn (@k13n)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/239
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-06-30
+---
+
+# Algorand Wallet Contextual KeyGen and Signing
+
+## Abstract
+
+This document specifies an expansion of the Algorand Wallet API to support contextual key generation and signing for **non-algorand-transaction** contexts.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative.
+
+### Overview
+
+> This overview section is non-normative.
+
+At a high level this specification defines a new API to generate keys and signing in other contexts than transactions.
+
+> Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects.
+
+#### Contexts
+
+An enum of contexts that are supported by the wallet. This enum is extensible and can be expanded in the future to support more contexts (i.e messaging, ephemeral keys, other chain derivations, etc)
+
+```ts
+enum KeyContext {
+    Address = 0, // Otherwise known as Account?
+    Identity = 1,
+    ...
+}
+```
+
+#### BIP44 Context Derivation Paths
+
+There is several reasons why we want to avoid key re-use in different contexts. One of the reasons is that we want to avoid the possibility of **leaking** information about the identity of the user. Another reason is that there is some cryptographic security concers when re-using keys or in case of keys being compromised the impact is **compartmentalized**.
+
+For this reason is advisable to have different derivation paths for different contexts supporting recoveribility of the keys by using the same mnemonic as before.
+
+| Purpose | Coin Type | Account | Change | Address Index |    Context    |
```

Comment:
> A few points here:
> 
> 1. Is there any particular reason why we opted for `purpose` 44 equal to BIP-44? IMO we should use a different `purpose` from Bitcoin's wallets.
> 2. Coin Type: we should provide default registered coin types (like [here](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#registered-coin-types)), at least for ALGO, and common ASAs e.g. USDC, gALGO, goETH, goBTC, ...
> 3. Are we enforcing the same account discovery algorithm ad BIP-44 here? Does it make sense to put a limit on account indexes that are valid only if previously active on-chain? What if I want to reserve 2/3 consecutive derivation paths for accounts that I will never use on-chain (maybe one for document encryption, another for authentication, erc..)

Hey Stefano,

So, regarding

1) We have discussed this before. It doesn't really matter that is the same path as in bitcoin, it's 2 different elliptic curves. And this is following the spec for the W3C Universal Wallet API(https://github.com/w3c-ccg/universal-wallet-interop-spec) 

2) Yea, we could extend it in the future to other ecosystems / applications

3) The discovery algorithm should be the same as any other chain wallet. Recursive finding until no account / address has any activity. For off-chain use, if you need to recovery, all is still recoverable, depending on the context you need to loop through them and see if you can use them, or can include them in wallet backups.

Reply:
1. Yes we use different curves. However, this ARC aims at providing "Contextual Wallet APIs" for different purposes and contexts. In that case, we might want to use another `purpose` as a derivation path. For example, [SLIP-0013](https://github.com/satoshilabs/slips/blob/ef6d7700cc/slip-0013.md) provides authentication for HD wallets and uses `purpose = 13'` for authentication with deterministic hierarchy.
2. Yes! Maybe providing a table with "reserved" values could be useful before having fragmented implementations in the ecosystem. WDYT?
3. It depends on the purpose and context. While I think the discovery algorithm is needed for multi-account HD wallets (BIP-44-like) it is not necessary in other cases like authentication.

Reply:
I don't understand the point about assigning the coin type to different ASAs. Given that Algo is needed to opt in and transact with an ASA regardless, how would assigning ASA IDs to coin type work? And what if you want/need to hold multiple ASAs in one address?

Reply:
What I like of different coin types for ASAs is that it force users to split their holdings into different addresses. This would improve security in case of compromised keys.

Maybe changing the coin type is not the best approach because of ALGO though.

As you mentioned, ALGO opt-in is a protocol requirement, so any address should be intended to hold ALGOs as well as another ASA. However, we could think of another derivation level dedicated to ASAs. Some related discussion [here](https://github.com/ethereum/EIPs/issues/85).

Reply:
I feel that the biggest risk here is that the seed gets compromised, in which case all addresses are compromised regardless of the derivation path. I think it is unlikely that a single derived private key (i.e. a scalar output of the deriveKey function) could be compromised without the seed or many addresses also being compromised in the process.

If we want users want to split their funds wallets can encourage them and make it easy to use different indices or different accounts for accepting different tokens.